File: solutions/074-search-a-2d-matrix/solution.py
from typing import List


class Solution:
    def search_matrix(self, matrix: List[List[int]], target: int) -> bool:
        m, n = len(matrix), len(matrix[0])
        row, col = 0, n - 1
        while row < m and col >= 0:
            if matrix[row][col] == target:
                return True
            elif matrix[row][col] > target:
                col -= 1
            else:
                row += 1
        return False


if __name__ == "__main__":
    sol = Solution()
    got = sol.search_matrix([[1, 3, 5, 7], [10, 11, 16, 20], [23, 30, 34, 60]], 3)
    status = "PASS" if got is True else "FAIL"
    print(f"{status} | => {got}")

File: solutions/074-search-a-2d-matrix/solution.go
package main

import "fmt"

func searchMatrix(matrix [][]int, target int) bool {
	//Z字型查找，从右上角开始查找，如果当前值大于目标值，则向左移动，如果当前值小于目标值，则向下移动
	m, n := len(matrix), len(matrix[0])
	row, col := 0, n-1
	for row < m && col >= 0 {
		if matrix[row][col] == target {
			return true
		} else if matrix[row][col] > target {
			col--
		} else {
			row++
		}
	}
	return false
}

func main() {
	got := searchMatrix([][]int{{1, 3, 5, 7}, {10, 11, 16, 20}, {23, 30, 34, 60}}, 3)
	fmt.Printf("PASS | %v\n", got)
}
